Martink rsa/loading data

martink-rsa opened the pull request:
## Loading the Data Feature (Fulfilled all AC):
1. Fetching data from the pushshift.io API.
2. A spinner shows while the data is being pulled and rotates every 1.5 seconds.

**Note:** I'm not sure about the basic way I've deducted a year from the current date (line 13 in `SearchPage.js`), and feel it should have been handled differently. I usually use the `moment` library when dealing with dates, but I've tried to abstain from using libraries in this project.

### Bug fix:
Fixed the anchor links to the two Info sections. I spent quite a few hours researching how to implement anchor links correctly in React and eventually settled with a library (`react-router-hash-link`) after not making much progress in handling it manually.

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Subreddit form

martink-rsa committed:
Loading data

martink-rsa committed:
Loading data

martink-rsa committed:
Merge branch 'martink-rsa/LoadingData' of https://github.com/ooloo-io/reddit-timer-martink-rsa into martink-rsa/LoadingData

martink-rsa committed:
Loading Data

martink-rsa commented:
Seems like I'm getting an error on the API request which was working earlier. Trying to resolve it.

martink-rsa committed:
Loading data

martink-rsa commented:
It appears that the issue isn't my side as the error has stopped occurring without changing the code. `Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at  Reason: CORS header 'Access-Control-Allow-Origin'`

jkettmann commented on package.json:
```diff
@@ -11,6 +11,7 @@
     "react": "^16.13.0",
     "react-dom": "^16.13.0",
     "react-router-dom": "^5.1.2",
+    "react-router-hash-link": "^1.2.2",
```
This ideally should have been a separate PR. It's always best to have one PR per ticket. Since the bugfix isn't related to any open PR you would have created a new branch from master and added the changes there. But it's OK now

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
-  async function fetchData() {
+  async function fetchData(searchTerm) {
+    const timeNow = new Date();
+    const timeLastYear = Math.round(timeNow.getTime() / 1000) - 31536000; // Seconds in year
```
I would define this as a constant outside of the SearchPage component. This way it's easy to understand how you calculate (seconds * minutes * hours * days) and the name is very clear.

```
const ONE_YEAR_IN_SECONDS = 60 * 60 * 24 * 365;
```

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
+  async function fetchData(searchTerm) {
+    const timeNow = new Date();
+    const timeLastYear = Math.round(timeNow.getTime() / 1000) - 31536000; // Seconds in year
     setIsLoading(true);
```
I would move `setIsLoading(true);` to `handleSearch`. Then you have a pure function here that doesn't have side effects. handleSearch has side effects already

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
-    return 0;
+    setIsLoading(false);
+    return 'Error';
   }
```
This is about `{isLoading ? <Spinner /> : null}`. You can write `{iseLoading && <Spinner />}` as well

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
     const response = await fetch(
-      'https://www.reddit.com/r/javascript/search.json?q=oop&limit=5',
+      `https://api.pushshift.io/reddit/search/submission/?subreddit=${searchTerm}&sort=desc&sort_type=created_utc&after=${timeLastYear}&size=500`,
       { mode: 'cors' },
```
Not sure if this is required!?

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
+      `https://api.pushshift.io/reddit/search/submission/?subreddit=${searchTerm}&sort=desc&sort_type=created_utc&after=${timeLastYear}&size=500`,
       { mode: 'cors' },
     );
     if (response.status === 200) {
```
Checking the status shouldn't be necessary. Afaik fetch should throw an error if something went wrong during the request. Simply `return resonse.json()` in any case.

That said you have to add a try catch block around in handleSearch

```

try {
  const data = await fetchData(searchTerm);
} catch (error) {
  ...
} finally {
  setIsLoading(false)
}
```

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
   const [isLoading, setIsLoading] = useState(false);
 
-  async function fetchData() {
+  async function fetchData(searchTerm) {
```
This function can be defined outside of the SearchPage component (still in the same file though). This makes it clear that it doesn't depend on any outside variables (pure function)

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -6,21 +6,22 @@ import Spinner from '../../components/Spinner/Spinner';
     setIsLoading(true);
     const response = await fetch(
-      'https://www.reddit.com/r/javascript/search.json?q=oop&limit=5',
+      `https://api.pushshift.io/reddit/search/submission/?subreddit=${searchTerm}&sort=desc&sort_type=created_utc&after=${timeLastYear}&size=500`,
```
Haha, you did a good research. I didn't find this API. I used `https://www.reddit.com/r/javascript/top.json` which is more complicated since you have to send 5 requests that depend on each other. So well done

martink-rsa committed:
Loading data

martink-rsa committed:
Loading data

martink-rsa commented:
I've made all the review changes.

I was unhappy with the way I had implemented everything so I refactored the code. My main concern was that I was adding the changes to the URL in previous features leading me to believe that the page should perform it's search whenever the URL changes.

I've now added a listener for when the URL changes, which will change the search parameters. A useEffect() with the search parameters as a dependency will then automatically trigger a new search.

Other changes include a search object containing parameters and a function that builds the API query string based on the keys in the search object.

I feel the refactor handles the project tasks more accurately, however I have an issue that I simply can't seem to resolve which is that I need to add my function `handleSearch` as a dependency. I can't seem to be able to do this properly. Please let me know if you don't like this approach, I'm fine with tossing the code if you prefer the previous way I was handling this feature.

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
+      sort: 'desc',
+      sort_type: 'created_utc',
+      // Current time in seconds minus one year
+      after: Math.round((new Date()).getTime() / 1000) - ONE_YEAR_IN_SECONDS,
```
I think you were asking about how to calculate this without Moment. This is totally fine. Alternatively, you could use [dayjs](https://github.com/iamkun/dayjs)

```
const after = dayjs()
  .subtract(1, 'year')
  .unix();
```

jkettmann commented on src/pages/SearchBar/SearchBar.js:
```diff
@@ -1,6 +1,6 @@
 import React, { useState } from 'react';
```
This and the style file should go to `src/pages/SearchPage` since the SearchBar component is not a page and belongs to the SearchPage

jkettmann commented on src/pages/SearchBar/SearchBar.js:
```diff
@@ -10,12 +10,11 @@ import {
   const [isValidSearch, setIsValidSearch] = useState(true);
-  const { handleSearch } = props;
 
   function handleSubmit(event) {
```
You can handle the update of the URL here. This simplifies the code IMO. You can remove the handlesSearch function here and inside the SearchPage

```
const history = useHistory();
const { subreddit } = useParams();

function handleSubmit(event) {
  event.preventDefault();
  if (isValidSearch) {
    history.push(`/search/${searchTerm}`);
  }
}
```

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
-import React, { useState } from 'react';
-import { useHistory } from 'react-router-dom';
+import React, { useState, useEffect } from 'react';
```
Here is the refactored component. I'll add some comments below to explain why I removed certain things. I think there was actually a misunderstanding. You don't need to have an extra URL listener since the component is re-rendered with every URL change already. That's because of the usage of `useParams`

```
import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import SearchPageWrapper from './SearchPage.style';
import SearchBar from '../SearchBar/SearchBar';
import Spinner from '../../components/Spinner/Spinner';

const ONE_YEAR_IN_SECONDS = 60 * 60 * 24 * 365;

async function handleSearch(subreddit) {
  const oneYearAgo = Math.round((new Date()).getTime() / 1000) - ONE_YEAR_IN_SECONDS;
  const url = `https://api.pushshift.io/reddit/search/submission/?subreddit=${subreddit}&sort=desc&sort_type=score&after=${oneYearAgo}&size=500`;

  const response = await fetch(url);
  return response.json();
}

function SearchPage() {
  const [isLoading, setIsLoading] = useState(true);
  const [searchResults, setSearchResults] = useState([]);
  const { subreddit } = useParams();

  // Search for new subreddit when parameters have changed
  useEffect(() => {
    setIsLoading(true);
    handleSearch(subreddit)
      .then((data) => setSearchResults(data))
      // TO-DO: Add notification to user that there was an error
      .catch((error) => console.error(error))
      .finally(() => setIsLoading(false));
  }, [subreddit]);

  console.log('searchResults', searchResults);

  return (
    <SearchPageWrapper>
      <SearchBar isLoading={isLoading} />
      {isLoading && <Spinner />}
    </SearchPageWrapper>
  );
}

export default SearchPage;
```

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
+}
+
 function SearchPage() {
   const history = useHistory();
```
This can be removed when we handle the URL in SearchBar

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
   const [isLoading, setIsLoading] = useState(false);
+  const [searchResults, setSearchResults] = useState([]);
+  const path = useParams().subreddit;
+  const [redditAPIObj, setRedditAPIObj] = useState({
```
I think keeping track of this reddit object overcomplicates things. `after` can be calculated just before the API call, so it doesn't need to be inside a state. The `subreddit` is actually already defined above as `path`. And the rest is constant.

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
   }
 
+  // Listen for if the URL has changed
+  useEffect(() => {
```
This useEffect is interesting. It is executed whenever `path` (= subreddit in the URL) changes. `history` is probably constant so doesn't have any effect.

When it runs a history listener is added that updates the reddit state object. But it actually uses the `path` that is returned from `useParams`.

In the end, this isn't doing anything :-). You can remove this `useEffect` and use the `subreddit` from `useParams` directly.

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
+  useEffect(() => {
+    handleSearch();
+    console.log(searchResults);
+  }, [redditAPIObj]);
```
This dependency has the same effect as using `path`. The `useEffect` will be executed whenever the URL changes.

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,38 +1,96 @@
+    return APILink;
+  }
+
+  async function handleSearch() {
```
I moved the `handleSearch` function out of the component. This way it's a constant and doesn't need to be added to the dependencies of `useEffect`. To make it a pure function though you need to move all external dependencies like `setIsLoading` and `setSearchResults` to the `useEffect`.

jkettmann reviewed (changes requested):
The direction with `useEffect` was already right. This can be confusing for sure, but with time, you'll have an eye for these kinds of optimizations

martink-rsa committed:
Loading data

martink-rsa commented:
Thank you for the review and refactor, that is a lot simpler. I see now where I could have greatly simplified it.

jkettmann commented on src/pages/SearchPage/SearchPage.js:
```diff
@@ -1,95 +1,40 @@
-async function fetchData(url) {
+async function handleSearch(subreddit) {
+  const oneYearAgo = Math.round((new Date()).getTime() / 1000) - ONE_YEAR_IN_SECONDS;
+  const url = `https://api.pushshift.io/reddit/search/submission/?subreddit=${subreddit}&sort=desc&sort_type=score&after=${oneYearAgo}&size=500`;
```
Btw: Not sure if you noticed but I replace the sorting by date with `score`

martink-rsa replied:
I did, thanks.
